### PR TITLE
add module build visual diffs

### PR DESCRIPTION
**Pull request by erwinmombay**
add a module build visual diff job using the `gulp serve --new_server` task which does a transform of the documents

### CODE REVIEW

**Review by rsimha — COMMENTED** [review]
I think a more maintainable way to do this is to make the job script `visual-diff-tests.js` take a command line arg, and parametrize the job definition in `config.yml`.

**Comment by danielrozenberg** [comment]
> I think a more maintainable way to do this is to make the job script `visual-diff-tests.js` take a command line arg, and parametrize the job definition in `config.yml`.

#25720 has an example of how to do it

File: build-system/tasks/visual-diff/index.js [hunk 1]
```diff
@@ -184,8 +184,9 @@ async function launchPercyAgent() {
 async function launchWebServer() {
   await startServer(
     {host: HOST, port: PORT},
-    {quiet: !argv.webserver_debug},
-    {compiled: true}
+    // eslint-disable-next-line google-camelcase/google-camelcase
+    {quiet: !argv.webserver_debug, new_server: argv.esm},
```
Comment (rsimha) [inline build-system/tasks/visual-diff/index.js:188]: Glad to see we're starting to use `new_server` in our tests!

File: build-system/tasks/serve.js [hunk 2]
```diff
@@ -46,6 +46,38 @@ const argv = minimist(process.argv.slice(2), {string: ['rtv']});
...
+
+/**
+ * @typedef {{
+ *   compiled: boolean,
+ *   esm: boolean,
+ *   cdn: boolean,
+ *   rtv: string?
+ * }}
+ */
+const ModeOptionsDef = {};
```
Comment (rsimha) [inline build-system/tasks/serve.js:79]: Very nice!

File: build-system/tasks/serve.js [hunk 3]
```diff
@@ -161,10 +193,11 @@ async function stopServer() {
 
 /**
  * Closes the existing server and restarts it
+ * @param {?serverOptionsDef} serverOptions
```
Comment (rsimha) [inline build-system/tasks/serve.js:196]: nit: uppercase S in `serverOptionsDef`

File: build-system/pr-check/visual-diff-tests.js [hunk 4]
```diff
@@ -30,21 +31,23 @@ const {runCiJob} = require('./ci-job');
 
 const jobName = 'visual-diff-tests.js';
 
+const esmFlag = argv.type === 'module' ? ' --esm' : '';
```
Comment (rsimha) [inline build-system/pr-check/visual-diff-tests.js:34]: For better readability and maintainability, can we replace this line with the following, and use `baseCommand` in the commands below?
```suggestion
let baseCommand = 'gulp visual-diff';
if (argv.type === 'module') {
  baseCommand += ' --esm';
}
```

File: .circleci/config.yml [hunk 5]
```diff
@@ -146,9 +146,14 @@ jobs:
       - setup_vm
       - install_chrome
       - run:
-          name: 'Visual Diff Tests'
-          command: node build-system/pr-check/visual-diff-tests.js
+          name: 'Visual Diff Tests (using << parameters.build_type >> build)'
```
Comment (rsimha) [inline .circleci/config.yml:149]: On looking at the CI logs, I think it's cleaner to just say `(<< parameters.build_type >>)` instead of `(using << parameters.build_type >> build)`.

Can we trim this name, and the names of the step under `Nomodule Tests`?

File: .circleci/config.yml [hunk 6]
```diff
@@ -146,9 +146,14 @@ jobs:
...
+          command: node build-system/pr-check/visual-diff-tests.js --type << parameters.build_type >>
```
Comment (rsimha) [inline .circleci/config.yml:150]: I see that you're using `parameters:` with `build_type` enum with an explicit command-line flag called `--type` which is consumed by `visual-diff-tests.js`. Meanwhile, `Nomodule Tests` is using `matrix:` and `parameters:` with `process.env.config`.

I definitely think we should standardize how we parametrize test jobs and use a single style in both places. FWIW, I think the explicit parameter here is more readable than the `process.env` method.

Would it be too much to ask you to make the change here? (If it's too messy, we can clean up in a separate PR.)

/cc @danielrozenberg

**Review by rsimha — COMMENTED** [review]
PR is looking great! Just a few comments below.

File: .circleci/config.yml (hunk 6 above)
Comment (danielrozenberg) [reply .circleci/config.yml:150]: 100% agree with you on that. Let me clean up `Module/Nomodule Tests` in a separate PR

**Comment by danielrozenberg** [comment]
Since both jobs use the same Percy write-only token, they'll both end up in the same project, and Percy will only report one as a status to GitHub - whichever of the two ends later would override the earlier.

Two possible solutions to this are:
1. create separate projects (`ampproject/amphtml-module` and `ampproject/amphtml-nomodule`) with their own tokens, passed as parameters in the job matrix
2. merge the two by running them in a single job by running the same test suite twice, on two executions of the server (with/without esm) and with adding `(module)`/`(nomodule)` to the individual snapshot names to differentiate between the two

I am _very slightly_ leaning towards option #1, but don't really care either way. @rsimha can you think of how to pass the obfuscated token to the PR jobs?

File: .circleci/config.yml (hunk 6 above)
Comment (erwinmombay) [reply .circleci/config.yml:150]: nope, willing to make the change! thanks for the recommendation

**Comment by rsimha** [comment]
> Since both jobs use the same Percy write-only token, they'll both end up in the same project, and Percy will only report one as a status to GitHub - whichever of the two ends later would override the earlier.

Oooh, excellent point, and good catch! TBH, after looking at the build logs on CircleCI and on Percy, I am leaning hard towards option 2. Here is my rationale:

- When you look at a Percy build for a single commit, it's useful to see all results (from module and nomodule) on one page. This way, if module breaks something but nomodule does not, it will be very obvious.
- Finalizing a visual build for a given commit from one CI job once and for all seems far simpler than trying to coordinate across two jobs, not to mention the cost of maintaining two tokens and two projects.
- Looking at the actual cost of parallelizing on CircleCI, VM setup / install takes ~2m, and running of all visual tests takes ~2m. I think it's better to test both build types from one job that runs for ~6m. (I assume it will work if we make the job wait for both the module and nomodule builds, download both, and run the correct kind of test server for each testing run.)

WDYT @erwinmombay @danielrozenberg? Maybe a little extra trouble up front to get this to work, but worth it in the end?

File: .circleci/config.yml (hunk 6 above)
Comment (erwinmombay) [reply .circleci/config.yml:150]: @rsimha sorry. i meant yes, i am willing to make the change :D sorry if i worded that in a confusing way

Comment (rsimha) [reply .circleci/config.yml:150]: Oh crap, it was my fault! I understood you just fine, but used the wrong smiley. What a week it has been :)

**Comment by danielrozenberg** [comment]
Again, no strong feelings either way. The overhead having two tokens/projects isn't that serious though, and we already discussed having separate projects for Chrome and for Firefox once Puppeteer fully supports the latter

**Comment by rsimha** [comment]
> Again, no strong feelings either way. The overhead having two tokens/projects isn't that serious though, and we already discussed having separate projects for Chrome and for Firefox once Puppeteer fully supports the latter

Okay, in that case, I'll lower my hard preference to a mild preference, and let @erwinmombay chime in since he's going to be the primary user of the Percy builds. 😃 

Re: adding new tokens to our CircleCI project, it's no trouble, and I can help do it. 👍

**Comment by erwinmombay** [comment]
@rsimha i think i would prefer option 2, just because seeing the results in a single page seems ideal

File: .circleci/config.yml [hunk 7]
```diff
@@ -193,7 +193,7 @@ jobs:
       - restore_karma_cache:
           cache_name: nomodule
       - run:
-          name: 'Nomodule Tests (using << parameters.config >>-config.json)'
+          name: 'Nomodule Tests (<< parameters.config >>-config.json)'
```
Comment (rsimha) [inline .circleci/config.yml:196]: Revert lines 196 and 217. (They're being fixed separately in #32922.)

File: .circleci/config.yml [hunk 8]
```diff
@@ -328,9 +328,11 @@ workflows:
       - 'Validator Tests':
           <<: *push_and_pr_builds
       - 'Visual Diff Tests':
-          <<: *push_and_pr_builds
+          name: 'Visual Diff Tests'
```
Comment (rsimha) [inline .circleci/config.yml:331]: nit: Remove this field and restore `<<: *push_and_pr_builds` to the top here.

File: build-system/pr-check/visual-diff-tests.js [hunk 9]
```diff
@@ -30,21 +30,25 @@ const {runCiJob} = require('./ci-job');
...
+  timedExecOrDie(`${baseCommand} --nobuild --master`);
 }
 
 function prBuildWorkflow() {
   process.env['PERCY_TOKEN'] = atob(process.env.PERCY_TOKEN_ENCODED);
   if (buildTargetsInclude(Targets.RUNTIME, Targets.VISUAL_DIFF)) {
     downloadNomoduleOutput();
     timedExecOrDie('gulp update-packages');
-    timedExecOrDie('gulp visual-diff --nobuild');
+    timedExecOrDie(`${baseCommand} --nobuild --esm`);
```
Comment (rsimha) [inline build-system/pr-check/visual-diff-tests.js:47]: Missing a call to `downloadModuleOutput()` after line 45, correct? (Same applies to line 36.)

File: build-system/pr-check/visual-diff-tests.js [hunk 10]
```diff
@@ -30,21 +30,25 @@ const {runCiJob} = require('./ci-job');
...
 const jobName = 'visual-diff-tests.js';
 
+const baseCommand = 'gulp visual-diff';
+
 function pushBuildWorkflow() {
   downloadNomoduleOutput();
   timedExecOrDie('gulp update-packages');
   process.env['PERCY_TOKEN'] = atob(process.env.PERCY_TOKEN_ENCODED);
-  timedExecOrDie('gulp visual-diff --nobuild --master');
+  timedExecOrDie(`${baseCommand} --nobuild --master`);
```
Comment (rsimha) [inline build-system/pr-check/visual-diff-tests.js:39]: This too will need to be duplicated with and without `--esm`.

File: build-system/pr-check/visual-diff-tests.js [hunk 11]
```diff
@@ -30,21 +30,25 @@ const {runCiJob} = require('./ci-job');
 
 const jobName = 'visual-diff-tests.js';
 
+const baseCommand = 'gulp visual-diff';
```
Comment (rsimha) [inline build-system/pr-check/visual-diff-tests.js:33]: This is no longer necessary. Cleaner to restore the full commands below.

File: build-system/tasks/serve.js [hunk 12]
```diff
@@ -193,19 +226,19 @@ async function serve() {
 
 /**
  * Starts a webserver at the repository root to serve built files.
- * @param {boolean=} lazyBuild
+ * @param {ServerOptionsDef=} serverOptions
  */
-async function doServe(lazyBuild = false) {
+async function doServe(serverOptions = {}) {
```
Comment (rsimha) [inline build-system/tasks/serve.js:231]: Don't forget to update this call-site:

https://github.com/ampproject/amphtml/blob/9d4bf3117c3ab3c63e964407e3a546723189294a/build-system/tasks/default-task.js#L57

File: build-system/tasks/visual-diff/index.js [hunk 13]
```diff
@@ -854,6 +856,7 @@ visualDiff.flags = {
   'webserver_debug': '  Prints debug info from the local gulp webserver',
   'percy_agent_debug': '  Prints debug info from the @percy/agent instance',
   'debug': '  Sets all debugging flags',
+  'esm': '  Serve ESM JS',
```
Comment (rsimha) [inline build-system/tasks/visual-diff/index.js:859]: Since this task is `visual-diff` and not `serve`, maybe reword this flag description to `Run tests against the module build`?

File: build-system/tasks/visual-diff/index.js [hunk 14]
```diff
@@ -388,6 +389,7 @@ async function generateSnapshots(webpages) {
   // no interactions, and each test that has in interactive tests file should
   // load those tests here.
   for (const webpage of webpages) {
+    webpage.name += argv.esm ? ' (Module)' : ' (Nomodule)';
```
Comment (rsimha) [inline build-system/tasks/visual-diff/index.js:392]: Glad this was just a one-line change. (Nice work by @danielrozenberg in writing this file to allow for it!)

File: build-system/tasks/serve.js [hunk 15]
```diff
@@ -193,19 +226,19 @@ async function serve() {
...
     await restartServer();
   };
   watch(serverFiles).on('change', debounce(watchFunc, watchDebounceDelay));
-  if (argv.new_server) {
+  if (argv.new_server || serverOptions.new_server) {
     buildNewServer();
   }
-  await startServer({}, {lazyBuild}, {});
-  if (lazyBuild) {
+  await startServer({}, {lazyBuild: serverOptions.lazyBuild}, {});
```
Comment (rsimha) [inline build-system/tasks/serve.js:240]: The third arg here (`ModeOptions`) will have to be `{esm: argv.esm}` for the server to actually serve the ESM binaries. We log the server mode here:

https://github.com/ampproject/amphtml/blob/9d4bf3117c3ab3c63e964407e3a546723189294a/build-system/server/app-utils.js#L59-L75

To locally test this, you can run the following:
```
gulp visual-diff --nobuild --esm --percy_disabled 
```

The server must print `Serving ESM JS`, and not `Serving minified JS`

**Review by rsimha — COMMENTED** [review]
Getting closer! A few more comments below, and two critical changes. (Right now, the PR is neither downloading ESM binaries nor serving them during testing 😃)

File: build-system/tasks/serve.js [hunk 16]
```diff
@@ -161,19 +204,12 @@ async function stopServer() {
...
+async function restartServer(serverOptions = {}) {
   stopServer();
-  if (argv.new_server) {
-    try {
-      buildNewServer();
-    } catch {
-      log(red('ERROR:'), 'Could not build', cyan('AMP Dev Server'));
-      return;
-    }
-  }
```
Comment (rsimha) [inline build-system/tasks/serve.js:174]: This is essential here. Why was this moved out? `doServe()` already builds the new server for the startup case.

File: build-system/tasks/serve.js [hunk 17]
```diff
@@ -81,15 +113,26 @@ function getMiddleware() {
...
+    'new_server' in serverOptions ? serverOptions.new_server : argv.new_server
+  ) {
+    try {
+      buildNewServer();
+    } catch {
+      log(red('ERROR:'), 'Could not build', cyan('AMP Dev Server'));
+      return;
+    }
+  }
+
```
Comment (rsimha) [inline build-system/tasks/serve.js:135]: Why was this moved here? `doServe()` already does this during normal startup.

File: build-system/tasks/serve.js [hunk 18]
```diff
@@ -161,19 +204,12 @@ async function stopServer() {
 
 /**
  * Closes the existing server and restarts it
+ * @param {ServerOptionsDef=} serverOptions
  */
-async function restartServer() {
+async function restartServer(serverOptions = {}) {
```
Comment (rsimha) [inline build-system/tasks/serve.js:209]: You're calling `restartServer()` on line 238 with `modeOptions`. Need to pass through the param here.

**Review by rsimha — COMMENTED** [review]
A few more comments that I think will simplify the code.

File: build-system/tasks/serve.js [hunk 19]
```diff
@@ -81,15 +113,26 @@ function getMiddleware() {
...
 async function startServer(
   connectOptions = {},
   serverOptions = {},
   modeOptions = {}
 ) {
+  if (
+    'new_server' in serverOptions ? serverOptions.new_server : argv.new_server
+  ) {
+    try {
+      buildNewServer();
```
Comment (rsimha) [inline build-system/tasks/serve.js:129]: I see that in #30681, you added a call to `buildNewServer()` to `integration.js`. I think that change should be reverted because the server startup logic should entirely live in `server.js`. Other tasks should trust the server to start itself correctly. Can we clean that up too?

File: build-system/tasks/serve.js (hunk 17 above)
Comment (erwinmombay) [reply build-system/tasks/serve.js:135]: sorry was just testing some stuff out originally on why it wasnt logging "serving esm", will revert.

File: build-system/tasks/serve.js (hunk 19 above)
Comment (erwinmombay) [reply build-system/tasks/serve.js]: agree with the statement, but `buildNewServer` isnt really start up logic and more compile logic. the code essentially isnt there (since its in typescript) before we make this call. `buildNewServer` i believe should be done before the task, i just placed it there since its the latest we can call it to make "lazy"

File: build-system/tasks/serve.js (hunk 16 above)
Comment (erwinmombay) [reply build-system/tasks/serve.js]: reverted

File: build-system/tasks/serve.js (hunk 18 above)
Comment (erwinmombay) [reply build-system/tasks/serve.js]: added the param. thanks!

File: build-system/tasks/serve.js (hunk 16 above)
Comment (erwinmombay) [reply build-system/tasks/serve.js]: so `doServe` does but the visual-diff code actually calls `startServer` bypassing doServe and the building of the js files. should i switch visualDiff to call `doServe` instead?

File: build-system/pr-check/visual-diff-tests.js [hunk 20]
```diff
@@ -31,19 +32,24 @@ const {runCiJob} = require('./ci-job');
...
   process.env['PERCY_TOKEN'] = atob(process.env.PERCY_TOKEN_ENCODED);
   if (buildTargetsInclude(Targets.RUNTIME, Targets.VISUAL_DIFF)) {
+    downloadModuleOutput();
     downloadNomoduleOutput();
     timedExecOrDie('gulp update-packages');
+    timedExecOrDie('gulp visual-diff --nobuild --esm');
     timedExecOrDie('gulp visual-diff --nobuild');
   } else {
+    timedExecOrDie('gulp visual-diff --empty --esm');
     timedExecOrDie('gulp visual-diff --empty');
```
Comment (danielrozenberg) [inline build-system/pr-check/visual-diff-tests.js:53]: So... this still generates two separate builds and has Percy report the 2nd one

Basically you have to loop through both module and nomodule `gulp serve` inside the visual-diff/index.js job - everything needs to run from when `launchPercyAgent()` is called and finished before `exitPercyAgent_()` - this is what starts and finalizes a build on Percy

File: build-system/tasks/serve.js (hunk 19 above)
Comment (erwinmombay) [reply build-system/tasks/serve.js]: do you think it would be imperative i make this change here, or can i follow up on a separate PR? I am working on it already but i wasn't sure if i could clean separate the issue out in another PR

File: build-system/tasks/serve.js (hunk 16 above)
Comment (erwinmombay) [reply build-system/tasks/serve.js]: i went ahead and moved the call to `buildNewServer` from `doServe` to `startServer`. LMK what you think

File: build-system/pr-check/visual-diff-tests.js (hunk 20 above)
Comment (erwinmombay) [reply build-system/pr-check/visual-diff-tests.js:53]: got it. ill make the change

File: build-system/tasks/serve.js (hunk 19 above)
Comment (rsimha) [reply build-system/tasks/serve.js:126]: #32891 makes this change and cleans up `integration.js`. Once reviewed and merged, we shouldn't need any of these changes.
